src/xdgdesktopportal.cpp

```diff
@@ -21,16 +21,30 @@
 
 #include "xdgdesktopportal.h"
 
+#ifdef WITH_KWAYLAND
+#include "waylandhelper.h"
+#endif
+
 #include <QDebug>
 #include <QWindow>
 #include <QX11Info>
 
-QString XdgDesktopPortal::parentWindow(const QWindow *activeWindow)
+QString XdgDesktopPortal::parentWindow(QWindow *activeWindow)
 {
     if (!QX11Info::isPlatformX11()) {
-        // TODO Implement Wayland window ID using https://wayland.app/protocols/xdg-foreign-unstable-v2
-        qWarning() << "Retrieving XDP window ID on Wayland not implemented";
+#ifdef WITH_KWAYLAND
+        auto *helper = new WaylandHelper(activeWindow);
+        QEventLoop loop;
+        QObject::connect(helper, &WaylandHelper::doneXdgExport, &loop, &QEventLoop::quit);
+        loop.exec();
+        auto handle = helper->exportedHandle();
+        if (handle.isEmpty()) {
+            return {};
+        }
```

Comment:
You don't need heap here:
```suggestion
        WaylandHelper wayland(activeWindow);
        QEventLoop loop;
        QObject::connect(&wayland, &WaylandHelper::doneXdgExport, &loop, &QEventLoop::quit);
        loop.exec();
        const QString handle = wayland.exportedHandle();
        if (handle.isEmpty()) {
            return {};
        }
```